README.md
<!-- Improved compatibility of back to top link: See: https://github.com/othneildrew/Best-README-Template/pull/73 -->
<a name="readme-top"></a>
<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
*** or simply open an issue with the tag "enhancement".
*** Don't forget to give the project a star!
*** Thanks again! Now go create something AMAZING! :D
-->



<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/CagriDK/AsterixEuroControl">
    <img src="images/logo.png" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">Euro Control Asterix Parser</h3>

  <p align="center">
    A Generic Euro Control Asterix Protocol Parser for Network Packages
    <br />
    <a href="https://github.com/CagriDK/AsterixEuroControl">View Demo</a>
    ·
    <a href="https://github.com/CagriDK/AsterixEuroControl/issues">Report Bug</a>
    ·
    <a href="https://github.com/CagriDK/AsterixEuroControl/issues">Request Feature</a>
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
    </li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>


<!-- ABOUT THE PROJECT -->
## About The Project

<pre>
Empty Fields of CAT34 Message shown below:
</pre>
[![Product Name Screen Shot1][product-screenshot1]](https://example.com)

<pre>
Using a binary message for test case :
ASTERIX data block at pos 0: cat=34; len=20
Data: 22 00 14 f6 00 02 02 41 6d ec 40 94 40 60 44 00 94 00 00 10
</pre>

<pre>
After resolving data the outcome of json files shown below (Page 1):
</pre>
[![Product Name Screen Shot2][product-screenshot2]](https://example.com)

<pre>
After resolving data the outcome of json files shown below (Page 2):
</pre>

[![Product Name Screen Shot3][product-screenshot3]](https://example.com)

<p align="right">(<a href="#readme-top">back to top</a>)</p>


<!-- ROADMAP -->
## Roadmap

- [x] Add CAT34 Message
- [x] Create a Test Message for CAT34
- [x] Create a Test Environment
- [x] Create a Manual Test Message Parser
- [x] Add generic template model for message parsing
- [x] Add a json handler for data storage and json writer for logging
- [x] Add CAT21 Message
- [ ] Add CAT48 Message
- [ ] Add CAT62 Message
- [x] Make the parser generic for all CAT Messages

See the [open issues](https://github.com/CagriDK/AsterixEuroControl/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->
## Contact

Çağrı Dükünlü - cagridknl@gmail.com

Project Link: [https://github.com/CagriDK/AsterixEuroControl](https://github.com/CagriDK/AsterixEuroControl)

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/othneildrew/Best-README-Template.svg?style=for-the-badge
[contributors-url]: https://github.com/othneildrew/Best-README-Template/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/othneildrew/Best-README-Template.svg?style=for-the-badge
[forks-url]: https://github.com/othneildrew/Best-README-Template/network/members
[stars-shield]: https://img.shields.io/github/stars/othneildrew/Best-README-Template.svg?style=for-the-badge
[stars-url]: https://github.com/othneildrew/Best-README-Template/stargazers
[issues-shield]: https://img.shields.io/github/issues/othneildrew/Best-README-Template.svg?style=for-the-badge
[issues-url]: https://github.com/othneildrew/Best-README-Template/issues
[license-shield]: https://img.shields.io/github/license/othneildrew/Best-README-Template.svg?style=for-the-badge
[license-url]: https://github.com/othneildrew/Best-README-Template/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/othneildrew
[product-screenshot1]: images/EmptyFields_34.PNG
[product-screenshot2]: images/FilledField1_34.PNG
[product-screenshot3]: images/FilledField2_34.PNG
[Next.js]: https://img.shields.io/badge/next.js-000000?style=for-the-badge&logo=nextdotjs&logoColor=white
[Next-url]: https://nextjs.org/
[React.js]: https://img.shields.io/badge/React-20232A?style=for-the-badge&logo=react&logoColor=61DAFB
[React-url]: https://reactjs.org/
[Vue.js]: https://img.shields.io/badge/Vue.js-35495E?style=for-the-badge&logo=vuedotjs&logoColor=4FC08D
[Vue-url]: https://vuejs.org/
[Angular.io]: https://img.shields.io/badge/Angular-DD0031?style=for-the-badge&logo=angular&logoColor=white
[Angular-url]: https://angular.io/
[Svelte.dev]: https://img.shields.io/badge/Svelte-4A4A55?style=for-the-badge&logo=svelte&logoColor=FF3E00
[Svelte-url]: https://svelte.dev/
[Laravel.com]: https://img.shields.io/badge/Laravel-FF2D20?style=for-the-badge&logo=laravel&logoColor=white
[Laravel-url]: https://laravel.com
[Bootstrap.com]: https://img.shields.io/badge/Bootstrap-563D7C?style=for-the-badge&logo=bootstrap&logoColor=white
[Bootstrap-url]: https://getbootstrap.com
[JQuery.com]: https://img.shields.io/badge/jQuery-0769AD?style=for-the-badge&logo=jquery&logoColor=white
[JQuery-url]: https://jquery.com 


CHANGELOG.md
# Changelog

## v1.0.0

### Initial Added or Changed
- Added README.md
- Added .gitignore
- Added MIT License file
- Fixed typos in templates
- Added table of contents
- Added checkboxes for major features on roadmap

### Added or Changed
- Added Template Models for CAT Messages
- Added message parsing methods to cat34 object
- Json file implementation added
- Added cat34 struct
- Added generic item parser class
- All fields are added for euro control asterix parsing

### Removed
- Some packages/libraries from acknowledgements I no longer use


main.cpp
#include <iostream>

#include <chrono>
#include <cstdlib>
#include <thread>
#include <memory>
#include <fstream>
#include <stdexcept>
#include <string>
#include <vector>
#include <cassert>
#include <algorithm>

#include "CATMessages/034/cat34.h"

int main(int, char **)
{
    nlohmann::json definition, mapping;
    definition = nlohmann::json::parse(std::ifstream("../CATMessages/034/cat034_1.26.json"));
    //definition = nlohmann::json::parse(std::ifstream("../CATMessages/034/cat021_0.26.json"));

    //      ASTERIX data block at pos 0: cat=34; len=20
    //      22 00 14 f6 00 02 02 41 6d ec 40 94 40 60 44 00 94 00 00 10
    
    //      ASTERIX data block at pos 0: cat=21; len=36
    //      15 00 24 fb 81 13 80 00 05 40 30 1c 29 3e 00 6e 4c c9 ff e6 06 d6 49 6c 2e 00 08 08 18 dc 31 ca 08 20 00 15

    const char *data = "\x22\x00\x14\xf6\x00\x02\x02\x41\x6d\xec\x40\x94\x40\x60\x44\x00\x94\x00\x00\x10";
    //const char *data = "\x15\x00\x24\xfb\x81\x13\x80\x00\x05\x40\x30\x1c\x29\x3e\x00\x6e\x4c\xc9\xff\xe6\x06\xd6\x49\x6c\x2e\x00\x08\x08\x18\xdc\x31\xca\x08\x20\x00\x15";

    std::cout << "CAT CATEGORY = " << static_cast<int>(static_cast<unsigned char>(data[0])) << "\n";

    CAT34 temp_Parser34(definition, data);

    CATMessageBase *temp_Parser = &temp_Parser34;
    temp_Parser->decodeData();
}
